qa/logger-usage/src/main/java/org/elasticsearch/qa/loggerusage/ESLoggerUsageChecker.java

```diff
@@ -0,0 +1,437 @@
+        private final String methodName;
+        private final String logMethodName;
+        private final int line;
+        private final String errorMessage;
+
+        public WrongLoggerUsage(String className, String methodName, String logMethodName, int line, String errorMessage) {
+            this.className = className;
+            this.methodName = methodName;
+            this.logMethodName = logMethodName;
+            this.line = line;
+            this.errorMessage = errorMessage;
+        }
+
+        @Override
+        public String toString() {
+            return "WrongLoggerUsage{" +
+                "className='" + className + '\'' +
+                ", methodName='" + methodName + '\'' +
+                ", logMethodName='" + logMethodName + '\'' +
+                ", line=" + line +
+                ", errorMessage='" + errorMessage + '\'' +
+                '}';
+        }
+
+        /**
+         * Returns an error message that has the form of stack traces emitted by {@link Throwable#printStackTrace}
+         */
+        public String getErrorLines() {
+            String fullClassName = Type.getObjectType(className).getClassName();
+            String simpleClassName = fullClassName.substring(fullClassName.lastIndexOf(".") + 1, fullClassName.length());
+            int innerClassIndex = simpleClassName.indexOf("$");
+            if (innerClassIndex > 0) {
+                simpleClassName = simpleClassName.substring(0, innerClassIndex);
+            }
+            simpleClassName = simpleClassName + ".java";
+            StringBuilder sb = new StringBuilder();
+            sb.append("Bad usage of ");
+            sb.append(LOGGER_CLASS).append("#").append(logMethodName);
+            sb.append(": ");
+            sb.append(errorMessage);
+            sb.append("\n\t");
+            sb.append(fullClassName);
+            sb.append(".");
+            sb.append(methodName);
+            sb.append("(");
+            sb.append(simpleClassName);
+            sb.append(":");
+            sb.append(line);
+            sb.append(")");
+            return sb.toString();
+        }
+    }
+
+    private static class ClassChecker extends ClassVisitor {
+        private String className;
+        private final Consumer<WrongLoggerUsage> wrongUsageCallback;
+        private final String methodName;
+
+        public ClassChecker(Consumer<WrongLoggerUsage> wrongUsageCallback, String methodName) {
+            super(Opcodes.ASM5);
+            this.wrongUsageCallback = wrongUsageCallback;
+            this.methodName = methodName;
+        }
+
+        @Override
+        public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+            this.className = name;
+        }
+
+        @Override
+        public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
+            if (methodName == null || methodName.equals(name)) {
```

Comment:
It **might** be more clear to replace methodName with `Predicate<String> methodsToCheck`. It would have helped me when I first scanned the code.